docs/typescript.md

```diff
@@ -2,13 +2,473 @@
+
+<!-- TODO: Add section for `@ts-expect-error` -->
+
+#### Avoid `any`
+
+`any` is the most dangerous form of explicit type declaration, and should be completely avoided if possible.
+
+- `any` doesn't represent the widest type, or indeed any type at all. `any` is a compiler directive for _disabling_ static type checking for the value or type to which it's assigned.
+- `any` suppresses all error messages about its assignee. This includes errors that are changed or newly introduced by alterations to the code. This makes `any` the cause of dangerous **silent failures**, where the code fails at runtime but the compiler does not provide any prior warning.
+- `any` subsumes all other types it comes into contact with. Any type that is in a union, intersection, is a property of, or has any other relationship with an `any` type or value is erased and becomes an `any` type itself.
+
+```typescript
+// Type of 'payload_0': 'any'
+const handler:
+  | ((payload_0: ComposableControllerState, payload_1: Patch[]) => void)
+  | ((payload_0: any, payload_1: Patch[]) => void);
+```
+<!-- TODO: Add more examples: For instance, if you have a function that is declared to return any which actually returns an object, all properties of that object will be any. -->
+
+- From this follows the most dangerous characteristic of `any`: it infects all surrounding and downstream code with its directive to suppress errors, expanding the surface area of code for which the TypeScript compiler can make no guarantees regarding type safety or runtime behavior.
+
+##### Try `unknown` and `never` instead
+
+###### `unknown`
+
+- `unknown` is the universal supertype i.e. the widest possible type, equivalent to the universal set(U).
+- Every type is assignable to `unknown`, but `unknown` is only assignable to `unknown`.
+- When typing the _assignee_, `any` and `unknown` are completely interchangeable since every type is assignable to both.
+- `any` usage is often motivated by a need to find a placeholder type that could be anything. `unknown` is the most likely type-safe substitute for `any` in these cases.
+
+###### `never`
+
+- `never` is the universal subtype i.e. the narrowest possible type, equivalent to the null set(∅).
+- `never` is assignable to every type, but the only type that is assignable to `never` is `never`.
+- When typing the _assigned_:
+  - `unknown` is unable to replace `any`, as `unknown` is only assignable to `unknown`.
+  - The type of the _assigned_ must be a subtype of the _assignee_.
+  - `never` is worth trying, as it is the universal subtype and assignable to all types.
+
+<!-- TODO: Add examples -->
+
+##### Don't allow `any` to be used as a generic default argument
+
+Some generic types use `any` as a default generic argument. This can silently introduce an `any` type into the code, causing unexpected behavior and suppressing useful errors.
+
+🚫
+
+```typescript
+const NETWORKS = new Map({
+  mainnet: '0x1',
+  goerli: '0x5',
+}); // Type 'Map<any, any>'
```

Comment:
Ah looks like `Map<any, any>` only happens if it's instantiated as an empty collection. Since this case is already covered in the "When instantiating an empty container type, provide a type annotation" entry, I'm good with only leaving the `jest.fn` example here. If we find any other good examples we can add them later.